raycast/BaseGame.py

```diff
@@ -0,0 +1,145 @@
+import collections
+import time
+from abc import ABCMeta
+
+import katagames_sdk.engine as kataen
+
+
+pygame = kataen.import_pygame()
+EventReceiver = kataen.EventReceiver
+EngineEvTypes = kataen.EngineEvTypes
+
+
+class BaseGame(metaclass=ABCMeta):
+    """
+    Base class for games, this is staged for the katasdk0.0.7 release
+    """
+
+    FPS_TRACKING_DEFAULT_SS = 13
+
+    def __init__(self, track_fps=True):
+        if track_fps:
+            self._fps_n_frames = self.FPS_TRACKING_DEFAULT_SS
+        else:
+            self._fps_n_frames = 0
+
+        self._fps_tracker_logic = collections.deque()
+        self._fps_tracker_rendering = collections.deque()
+        self._tick = 0
+
+        self._cached_info_text = None
+        self._info_font = None
+
+    def start(self):
+        """Starts the game loop. This method will not exit until the game has finished execution."""
+
+        class _ProxyReceiver(EventReceiver):
+            def __init__(self, ref_game):
+                super().__init__()
+                self._tnow_cache = None
+                self._game = ref_game
+                self._event_queue = []
+                self._last_update_time = time.time()
+
+            def proc_event(self, ev, source):
+                if ev.type == EngineEvTypes.LOGICUPDATE:
+                    self._game._update_internal(self._event_queue, ev.curr_t, ev.curr_t-self._last_update_time)
+                    self._last_update_time = self._tnow_cache = ev.curr_t
+                    self._event_queue.clear()
+                if ev.type == EngineEvTypes.PAINT:
+                    self._game._render_internal(ev.screen, self._tnow_cache)
+                else:
+                    self._event_queue.append(ev)
+
+        kataen.init(self.__get_mode_internal())
+        b, a = kataen.get_game_ctrl(), _ProxyReceiver(self)
+        a.turn_on()
+        b.turn_on()
+
+        self.pre_update()
+        b.loop()
+        kataen.cleanup()
+
+    """
+    implement methods below so the game class suits your needs
+    """
+    def get_mode(self) -> str:
+        """
+        you need to return either
+        'HD', 'OLD_SCHOOL', or 'SUPER_RETRO'
+        """
+        raise NotImplementedError()
```

Comment:
This is more of a question than a comment, but is there any reason why games can't specify an exact resolution / scaling factor? It would be nice if you could just tell the screen to be WxH, at a scaling factor of S, for example. If that were possible, this method could be redone to not use magic strings at all (which I personally don't like)

Reply:
It's a long story… If you're interested you can read about it in the [updated FAQ(developers doc)](https://kata.games/developers/faq)